Replace deprecated BERT library with HuggingFace

nagyrajmund opened the pull request:
This PR replaces the deprecated `bert-embedding` library with the BERT implementation from HuggingFace. 

It also includes a short script for verifying that the embedding vectors are the same as before: on an example sentence, the maximum of the element-wise difference between word embeddings of the old and the new implementation is on the order of 1e-6.

[WIP]: My laptop fails with a `MemoryError` when trying to save the text data into files at the end of the preprocessing script. However, the shape of the vector seems correct so I'm not sure what's the issue.

[WIP]: I haven't verified yet that the outputs of the new code are the same as the old code.

nagyrajmund committed:
Replace deprecated BERT library with HuggingFace

nagyrajmund commented:
Don't forget to run `pip install transformers[torch]` before testing the changes.

nagyrajmund commented on gesticulator/data_processing/process_dataset.py:
```diff
@@ -97,7 +98,11 @@ def _encode_vectors(audio_filename, gesture_filename, text_filename, embedding_m
     # Step 3: Obtain text transcription:
     if isinstance(embedding_model, BertEmbedding):
-        text_encoding = encode_json_transcript_with_bert(text_filename, embedding_model)
+        print("[WARNING] Using deprecated BERT model!")
+        text_encoding = encode_json_transcript_with_bert_DEPRECATED(text_filename, embedding_model)
```
I left in the old functionality for now to facilitate the testing of the code.

nagyrajmund commented on gesticulator/data_processing/process_dataset.py:
```diff
@@ -232,6 +237,11 @@ def _save_dataset(data_csv, save_dir, embedding_model, dataset_name, args):
+        bert_model = BertModel.from_pretrained('bert-base-cased')
+
+        return tokenizer, bert_model
+    if name == "BERT_deprecated":
```
The old implementation can be enabled with this parameter

nagyrajmund commented on gesticulator/data_processing/text_features/parse_json_transcript.py:
```diff
@@ -289,3 +291,177 @@ def extract_extra_features(word_attributes, total_elapsed_time):
+
+# -----------------------------------------------------------------
+
+def encode_json_transcript_with_bert_DEPRECATED(json_file, bert_model):
```
This is the old implementation, I just added _DEPRECATED to the end

Svito-zar commented:
Yeah, it would be great to address issue https://github.com/Svito-zar/gesticulator/issues/13 , but there is no rush and we could finish it after the holidays as well

clarksandholtz commented:
Would love to see this get changed to HuggingFace. Having a hard time with `bert-embedding` and it's `mxnet` dependency. 

Is this still something you are planning on doing?

Svito-zar commented:
Yes, @clarksandholtz, we plan to change to HuggingFace in order to ease all the dependencies.
@nagyrajmund is going to finish this pull request within a few weeks

nagyrajmund committed:
Merge branch 'master' of github.com:Svito-zar/gesticulator

nagyrajmund committed:
Use os.path.join() instead of path concatenation
in rename_data_files.py

nagyrajmund committed:
Add default folder args to bvh2features.py

nagyrajmund committed:
Fix a bug in text processing that caused out-of-memory crashes

nagyrajmund committed:
BUGFIX: add missing BERT tokenization alignment

The JSON transcription of the speech has a different idea of what a "word" is than BERT, e.g. BERT splits long words into multiple subwords. These subwords needed to be merged into one to keep the correspondence between words in the JSON and words in the BERT encoding.

nagyrajmund committed:
Add a command to the README for downloading the GENEA dataset

nagyrajmund committed:
Update README.md

nagyrajmund committed:
Remove old BERT library from requirements

nagyrajmund committed:
Update requirements.txt

nagyrajmund commented on gesticulator/data_processing/text_features/parse_json_transcript.py:
```diff
@@ -87,10 +86,14 @@ def encode_json_transcript_with_bert(json_file, tokenizer, bert_model):
+                # If filler is the last word in the sentence, then we add its delimiter to the last word in the sentence
                 if word[-1] in delimiters: 
-                    non_filler_words_in_sentence.append(word[-1])
+                    if non_filler_words_in_sentence != []:
+                        # Append to the last word
+                        non_filler_words_in_sentence[-1] += word[-1]
+                    else:
+                        # Add the delimiter as the only word in the sentence
+                        non_filler_words_in_sentence.append(word[-1])
```
this if was added in case the sentence is empty (e.g. "Ah." will be ["."] since "Ah" is ignored"

Svito-zar commented:
@clarksandholtz , this PR is finished and have been just merged to the repo.
Thank you @nagyrajmund for your hard work

clarksandholtz commented:
Thank you @nagyrajmund for implementing change! So great!

Thank you @Svito-zar for the update.

This is an awesome model and I'm really excited to play around with it.
